content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+differ).
+
+Then follow the setup instructions
+[here](https://mlir.llvm.org/docs/Tools/MLIRLSP/#td---tablegen-files) to tell
+the extension where the server and compilation database are.
+
+## Example
+
+Assuming you configured LLVM with the AArch64 target enabled (it is on by
+default).
+
+* Open the file `llvm/lib/Target/AArch64/AArch64.td`.
+* Put your cursor on a use of the `SubtargetFeature` type.
+* In the top menus select "Go" then "Go to Definition".
+* You will be taken to `llvm/include/llvm/Target/Target.td`, where
+  `SubtargetFeature` is defined.
+
+## Limitations
+
+The language server highlights an anti-pattern in the way some LLVM targets
+(e.g. AArch64) use TableGen.
+
+You may find yourself in a file that uses a class but does not define it, or
+include any files which do. This is because this file is intended to be included
+in another file, which does include a definition of that class.
+
+Perhaps we can address this by improving the language server, or reorganising
+the includes so we do not have these seemingly isolated files.
+
+# Dump
+
+What about printf? The best debugging tool of them all.
+
+TableGen’s equivalent is
+[dump](https://llvm.org/docs/TableGen/ProgRef.html#dump-print-messages-to-stderr),
+and its companion `repr`.
+```
+class Register<int _size, string _alias=""> {
+   int size = _size;
+   string alias = _alias;
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+}
+
+def X0: Register<8> {}
+def X29: Register<8, "frame pointer"> {}
+```
+
+`dump` prints to `stderr`:
+```
+<source>:4:5: note: size is 8 and alias is ""
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+   ^
+<source>:4:5: note: size is 8 and alias is "frame pointer"
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+   ^
+```
+
+Note: This was added
+[recently](https://github.com/llvm/llvm-project/commit/411c4edeef076bd2e01b104fe095ba381600a3d3)
+by Francesco Petrogalli and Adam Nemet. So you will need a recent build, or a
+released version 18.0 or greater (which is unreleased at time of writing).
+
+Of course you can try this
+[on Compiler Explorer](https://godbolt.org/z/555n19qMK) right now!
+
+# Find In Files
+
+This is last because in an ideal world it would be the last option, but it’s
+often not the least of them. Grep, ack, Find In Files, whatever you call it,
+searching text is unreasonably effective if you have a little knowledge of the
+language syntax.
+
+Should I even mention such an obvious idea? Well obvious is in the eye of the
+beholder, and for TableGen we have a special circumstance that makes it more
+effective than usual.
+
+In the LLVM project repo we have the vast majority of TableGen code in use today.
+Want to know how to use a feature? It’s all there, albeit in a 500,000+ line
+haystack, but you’d be surprised what a simple query can find.
+
+Think about how you would write the thing you’re trying to find. If it’s a
```

Comment:
write -> describe?

Reply:
So I said "write" because the idea is if you wanted to find `class Foo`, you'd think about how you'd write the source code to describe that class.

And then I realised I should write it like that instead. Hope that it is clearer now.

(describe to me sounds like a more "intelligent" search that knows more than just the raw text)